### DISCUSSION TITLE
The return type from a model operation such as findMany is incorrect with conditionally passed arguments

### DISCUSSION THREAD

**Discussion by sambartik** [prisma/prisma#20725] [Q&A]
### Question

Hello, I have a piece of code that is a bit complicated, so I simplified it a bit to make it easier to reproduce, however I think this still demonstrates the issue well enough. I wanted to know whether you agree with me that this is in fact a bug before filing a new issue.

I have cloned one of the prisma example repos so the prisma scheme is almost the same - there are 2 models: `User` and `Post`. In my code there is a boolean variable `includePosts` based on which the arguments to `findMany` operation are built:

```ts
const users = await prisma.user.findMany(
  includePosts
    ? {
        include: {
          posts: true,
        },
      }
    : undefined,
);
```

The problem arises with the type of the `users` variable. In this case, the type indicates that it ALWAYS includes the `posts` property, which is not true. It depends based on the value of `includePosts` variable. Therefore typescript does not rise any errors in the following code block:

```ts
  if (includePosts) {
    console.log("Posts are included in the users prisma payload.");
    users.forEach(user => {
      console.log(user.posts);
    });
  } else {
    console.log("Posts are NOT included in the users prisma payload.");
    users.forEach(user => {
      console.log(user.posts); // <------ This should be a type error
    })
  }
```

What do you think? You can find the source code of the example over [here](https://github.com/sambartik/prisma-example/). I believe I have also found a slightly different variation of this problem, which I have published in [another branch of the same repo](https://github.com/sambartik/prisma-example/tree/variation-2).

### How to reproduce (optional)

1. `git clone https://github.com/sambartik/prisma-example.git`
2. `npm i`
3. `npx prisma migrate dev`
4. `npm run dev`

### Expected behavior (optional)

Typescript should throw a type error

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String
  posts Post[]
}

model Post {
  id        Int      @id @default(autoincrement())
  title     String
  content   String
  published Boolean  @default(false)
  author    User    @relation(fields: [authorId], references: [id])
  authorId  Int
}

```

- OS: Windows
- Database: SQLite
- Node.js version: v18.17.0

- Run `prisma -v` to see your Prisma version and paste it
```
prisma                  : 5.1.1
@prisma/client          : 5.1.1
Current platform        : windows
Query Engine (Node-API) : libquery-engine 6a3747c37ff169c90047725a05a6ef02e32ac97e (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Schema Engine           : schema-engine-cli 6a3747c37ff169c90047725a05a6ef02e32ac97e (at node_modules\@prisma\engines\schema-engine-windows.exe)
Schema Wasm             : @prisma/prisma-schema-wasm 5.1.1-1.6a3747c37ff169c90047725a05a6ef02e32ac97e
Default Engines Hash    : 6a3747c37ff169c90047725a05a6ef02e32ac97e
Studio                  : 0.492.0
```

**Comment by nurul3101** [comment]
Hello @sambartik 👋 

Thank you for sharing the detailed description.

I tried reproducing the issue with the sample schema that you provided and was able to find a solution that would work.

```ts
import { PrismaClient, Prisma } from "@prisma/client";

const prisma = new PrismaClient();

async function main() {
  let includePosts = false;

  if (includePosts) {
    const users: Prisma.UserGetPayload<{ include: { posts: true } }>[] =
      await prisma.user.findMany({
        include: {
          posts: true,
        },
      });

    console.log("Posts are included in the users prisma payload.");
    users.forEach((user) => {
      console.log(user.posts); // TypeScript knows that posts are included
    });
  } else {
    const users: Prisma.UserGetPayload<{}>[] = await prisma.user.findMany();

    console.log("Posts are NOT included in the users prisma payload.");
    users.forEach((user) => {
      // TypeScript should raise a type error here
      console.log(user.posts);
    });
  }
}

main()
  .catch((e) => {
    throw e;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
```
In this case, TypeScript correctly throws the error when `posts` are not fetched.
Can you try this and check if it works for you?

<img width="1420" alt="Screenshot 2023-08-17 at 10 08 12 PM" src="https://github.com/prisma/prisma/assets/13049676/f872027e-0b44-405d-83ad-81605d493c7b">

**Reply by sambartik** [reply]
Hi! Thank you for such quick answer! 

This is essentially what I ended up doing in my own project, however this feels very much as a hack and in my case more importantly it violates the DRY principle, because I have arguments that both Payloads (the one with and without the posts relation) share. 

Because of that, it opens me vulnerable to a possible bug in the future, where I would in the time of need forget to change the shared arguments in both places (in your photo the lines number 9 & 21) and update accordingly in both places.

**Reply by sambartik** [reply]
To elaborate on my last point, I managed to find what I think is a good compromise. The only downside is that it uses more memory.

```ts
  const sharedArgs: Prisma.UserFindManyArgs = {};
  const withoutPostsArgs = {
    ...sharedArgs,
    // Could also add additional arguments, special when without posts...
    // .
    // .
  } satisfies Prisma.UserFindManyArgs;
  const withPostsArgs = {
    ...sharedArgs,
    include: {
      posts: true,
    },
    // Could add another additional arguments, special when with posts...
    // .
    // .
  } satisfies Prisma.UserFindManyArgs;

  const users: (
    | Prisma.UserGetPayload<typeof withPostsArgs>
    | Prisma.UserGetPayload<typeof withoutPostsArgs>
  )[] = await prisma.user.findMany(includePosts ? withPostsArgs : withoutPostsArgs);

  users.forEach((user) => {
    if ("posts" in user) {
      console.log(user.posts);
    } else {
      console.log(user.posts); // This shows correctly as a type error
    }
  });
``` 
[(src)](https://github.com/sambartik/prisma-example/tree/workaround)

After some additional research, I don't feel like this is a problem at the prism side of things, but it would have been helpful to add a mention in the docs to be aware of something like this.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
